### PR TITLE
New Utf8.size() method to get a string's UTF-8 length.

### PR DESCRIPTION
**Pull request by swankjesse**
Closes: https://github.com/square/okio/issues/283

### CODE REVIEW

File: okio/src/main/java/okio/Buffer.java [hunk 1]
```diff
@@ -843,7 +843,7 @@ public void clear() {
 
   @Override public Buffer writeUtf8(String string, int beginIndex, int endIndex) {
     if (string == null) throw new IllegalArgumentException("string == null");
-    if (beginIndex < 0) throw new IllegalAccessError("beginIndex < 0: " + beginIndex);
+    if (beginIndex < 0) throw new IllegalArgumentException("beginIndex < 0: " + beginIndex);
```
Comment (swankjesse) [inline okio/src/main/java/okio/Buffer.java]: `IAE<tab>`

File: okio/src/main/java/okio/Utf8.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okio;
+
+/** APIs for working with UTF-8 strings. */
+public final class Utf8 {
```
Comment (swankjesse) [inline okio/src/main/java/okio/Utf8.java]: Would it be simpler as just `Okio.utf8Size()` ? I think this is pretty discoverable

File: okio/src/main/java/okio/Utf8.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  private Utf8() {
+  }
+
+  /**
+   * Returns the number of bytes used to encode {@code string} as UTF-8 when using {@link
+   * ByteString#encodeUtf8} or {@link Buffer#writeUtf8(String)}.
```
Comment (cketti) [inline okio/src/main/java/okio/Utf8.java]: `ByteString#encodeUtf8()` currently uses `String#getBytes(Charset)`. Is the behavior of `getBytes()` with UTF-8 as charset sufficiently documented to guarantee this method exhibits the same behavior as Java runtime implementations?

Comment (swankjesse) [inline okio/src/main/java/okio/Utf8.java]: Good question. We’ve got test cases for a bunch of interesting cases, and we strive to be compatible. Could probably use even more tests though – I know runs of mismatched surrogates have differed between Dalvik and Android in at least one case.

Comment (cketti) [inline okio/src/main/java/okio/Utf8.java]: Sounds like the `ByteString#encodeUtf8()` part should be dropped from the JavaDoc. Alternatively, `ByteString#encodeUtf8()` could be changed to manually UTF-8-encode the input string.

Comment (swankjesse) [inline okio/src/main/java/okio/Utf8.java]: What about just testing to confirm it’s consistent? The JDK’s implementation is faster, but only when it needs to allocate a `byte[]` anyway.

File: okio/src/main/java/okio/Utf8.java (hunk 2 above)
Comment (JakeWharton) [inline okio/src/main/java/okio/Utf8.java]: My only (very minor) complaint about having a `Utf8` class is that none of the other operations dealing with UTF are exposed here.

Comment (swankjesse) [inline okio/src/main/java/okio/Utf8.java]: Lemme add Javadoc referencing all the other UTF-8 things

**Comment by swankjesse** [comment]
screen shot 2017-04-11 at 10 23 29 pm

With apologies for bad default CSS.

**Pull request merged**